src/gtc/passes/oir_optimizations/vertical_loop_merging.py

```diff
@@ -0,0 +1,64 @@
+# -*- coding: utf-8 -*-
+#
+# GTC Toolchain - GT4Py Project - GridTools Framework
+#
+# Copyright (c) 2014-2021, ETH Zurich
+# All rights reserved.
+#
+# This file is part of the GT4Py project and the GridTools framework.
+# GT4Py is free software: you can redistribute it and/or modify it under
+# the terms of the GNU General Public License as published by the
+# Free Software Foundation, either version 3 of the License, or any later
+# version. See the LICENSE.txt file at the top-level directory of this
+# distribution for a copy of the license or check <https://www.gnu.org/licenses/>.
+#
+# SPDX-License-Identifier: GPL-3.0-or-later
+
+from typing import Any
+
+from eve import NodeTranslator
+from gtc import common, oir
+
+
+class AdjacentLoopMerging(NodeTranslator):
+    @staticmethod
+    def _mergeable(a: oir.VerticalLoop, b: oir.VerticalLoop) -> bool:
+        if a.loop_order != b.loop_order:
+            return False
+        a_interval = a.sections[-1].interval
+        b_interval = b.sections[0].interval
+        if a.loop_order == common.LoopOrder.BACKWARD:
+            a_lim = a_interval.start
+            b_lim = b_interval.end
+        else:
+            a_lim = a_interval.end
+            b_lim = b_interval.start
+        return a_lim.level == b_lim.level and a_lim.offset == b_lim.offset
+
+    @staticmethod
+    def _merge(a: oir.VerticalLoop, b: oir.VerticalLoop) -> oir.VerticalLoop:
+        sections = a.sections + b.sections
+        return oir.VerticalLoop(
+            loop_order=a.loop_order,
+            sections=sections,
+            caches=[],  # TODO: add support for cache merging?
```

Comment:
regarding the TODO: probably the merge should run before the cache optimizations? If yes, then we could warn or assert here if caches in a.caches and b.caches are not empty.

Reply:
Replaced the TODO by a warning in case there are removed caches. The pass should indeed run before cache detection and that’s also what’s implemented in the current GTC backends.